src/Graph.ts

```diff
@@ -454,6 +456,15 @@ export default class Graph {
 			module.dependencies.forEach(depModule => {
 				if (!depModule.isExternal) {
 					visit(<Module>depModule, seen);
+					const depId = relativeId(depModule.id);
+					cycles.forEach(cycle => {
+						if (cycle.path[depId] === null) {
+							cycle.path[depId] = modId;
+							if (cycle.id !== modId) {
+								cycle.path[modId] = null;
+							}
```

Comment:
I will have to think about this further as I'm not sure I understand all the edge cases. As I say I'd be much happier merging if we can move as much of this logic into the warning function itself, with simple non-loop-based analysis passed through rather (eg having the `seen` object provide a parent map or similar information that doesn't involve extra looping to create).

Otherwise I'll want to test this much more thoroughly for edge cases and loop complexity.